internal/users/db/db.go

```diff
@@ -0,0 +1,256 @@
+// Package db handles transaction with an underlying database to store user and group information.
+package db
+
+import (
+	"context"
+	"database/sql"
+	_ "embed"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+
+	// sqlite3 driver.
+	_ "github.com/mattn/go-sqlite3"
+	"github.com/ubuntu/authd/internal/fileutils"
+	"github.com/ubuntu/authd/internal/users/db/bbolt"
+	"github.com/ubuntu/authd/log"
+)
+
+var (
+	filename = "authd.sqlite3"
+	//go:embed sql/create_schema.sql
+	createSchema string
+)
+
+// Database is our database API.
+type Database struct {
+	db   *sql.DB
+	path string
+	mu   sync.RWMutex
+}
+
+// GroupDB is the struct stored in json format in the database.
+type GroupDB struct {
+	Name  string
+	GID   uint32
+	UGID  string
+	Users []string
+}
+
+// queryable is an interface to execute SQL queries. Both sql.DB and sql.Tx implement this interface.
+type queryable interface {
+	Exec(query string, args ...interface{}) (sql.Result, error)
+	QueryRow(query string, args ...any) *sql.Row
+	Query(query string, args ...any) (*sql.Rows, error)
+}
+
+// New creates a new database by creating or opening the underlying db.
+func New(dbDir string) (*Database, error) {
+	dbPath := filepath.Join(dbDir, filename)
+
+	exists, err := fileutils.FileExists(dbPath)
+	if err != nil {
+		return nil, err
+	}
+	if !exists {
+		// Create the database with permissions 0600.
+		if err := fileutils.Touch(dbPath); err != nil {
+			return nil, err
+		}
+	}
+
+	// Fail if permissions are not 0600
+	// TODO: I don't see why we should fail here instead of just fixing the permissions.
+	fileInfo, err := os.Stat(dbPath)
+	if err != nil {
+		return nil, fmt.Errorf("can't stat database file: %v", err)
+	}
+	perm := fileInfo.Mode().Perm()
+	if perm != 0600 {
+		return nil, fmt.Errorf("wrong file permission for %s: %o", dbPath, perm)
+	}
```

Comment:
To the reviewer: I copied this from the old code but added this TODO comment:
```
	// TODO: I don't see why we should fail here instead of just fixing the permissions.
```

Is there a good reason for failing or should we just fix the permissions?

Reply:
I think it's because, if the permissions aren't correct, we can't vouch for the database integrity anymore, so that's why we would fail here (yeah, root users can always edit it, but if others can read or write, the database isn't safe anymore).

Reply:
If that's the rationale, then I think we should instead check if the file is writable to group or others instead. Because the way it is now, it also fails in cases which do not allow unauthorized manipulation, for example if the file is read-only (0400) or readable to group or others (0644).

Reply:
I agree with that. Now our cache only represents the passwd database, IIRC, so we should check for `644` (we checked for ` 600` for the shadow database, since only root has access to it).

Reply:
I pushed a commit for that:

    Check database owner and permissions
    
    ... but don't fail if the database has other permissions than 0600,
    because it doesn't contain any secrets, so it's fine if it's readable by
    other users than the owner. Check that it's not *writable* by other
    users though.